algorithms/linfa-elasticnet/src/algorithm.rs

```diff
@@ -187,6 +333,86 @@ fn coordinate_descent<'a, F: Float>(
     (w, gap, n_steps)
 }
 
+fn block_coordinate_descent<'a, F: Float>(
+    x: ArrayView2<'a, F>,
+    y: ArrayView2<'a, F>,
+    tol: F,
+    max_steps: u32,
+    l1_ratio: F,
+    penalty: F,
+) -> (Array2<F>, F, u32) {
+    let n_samples = F::cast(x.shape()[0]);
+    let n_features = x.shape()[1];
+    let n_tasks = y.shape()[1];
+    // the parameters of the model
+    let mut w = Array2::<F>::zeros((n_features, n_tasks));
+    // the residuals: `Y - XW` (since W=0, this is just `Y` for now),
+    // the residuals are updated during the algorithm as the parameters change
+    let mut r = y.to_owned();
+    let mut n_steps = 0u32;
+    let norm_cols_x = x.map_axis(Axis(0), |col| col.dot(&col));
+    let mut gap = F::one() + tol;
+    let d_w_tol = tol;
+    let tol = tol * y.map(|yij| yij.powi(2)).sum();
+    while n_steps < max_steps {
+        let mut w_max = F::zero();
+        let mut d_w_max = F::zero();
+        for j in 0..n_features {
+            if abs_diff_eq!(norm_cols_x[j], F::zero()) {
+                continue;
+            }
+            let old_w_j: ArrayView1<F> = w.slice(s![j, ..]);
+            let x_j: ArrayView1<F> = x.slice(s![.., j]);
+            let norm_old_w_j = old_w_j.map(|wjt| wjt.powi(2)).sum().sqrt();
+            if abs_diff_ne!(norm_old_w_j, F::zero()) {
+                for i in 0..x.shape()[0] {
+                    for t in 0..n_tasks {
+                        r[[i, t]] += x_j[i] * old_w_j[t];
+                    }
+                }
```

Comment:
Isn't this operation equivalent to `r += x_j.dot(old_w_j.t())`? If so you can replace these types of for loops with `general_mat_mul(1, x_j, old_w_j.t(), 1, r)`.

Reply:
This operation is equivalent to `np.outer` in Python (see: https://numpy.org/doc/stable/reference/generated/numpy.outer.html for a more detailed explanation). I don't think there is a built-in equivalent in ndarray. It is not yet available in the `ndarray` crate (see: https://github.com/rust-ndarray/ndarray/issues/1148) .

Reply:
`outer(x, w)` is equivalent to `x` as a column vector multiplied with `w` as a row vector. The code in the `ndarray` PR you linked does the same thing. Using `general_mat_mul` allows you to add the matrix product of `x` and `w` to `r` in one operation. You do need to convert `x` and `y` into 2D arrays though, like in the PR.

Reply:
Ideal way to convert 1D arrays into 2D is `insert_axis`. Something like `x.view().insert_axis(Axis(0_or_1))`